File: Content/iOS/Pascal Interperter/Swift Version/pascal09/pascal09/Interpreter.swift
//
//  Interpreter.swift
//  12312
//
//  Created by pmst on 2018/4/15.
//  Copyright © 2018 pmst. All rights reserved.
//

import Foundation


class Interpreter {
    var parser:Parser
    var global_scope :[String:Int] = [:]
    
    init(parser:Parser) {
        self.parser = parser
    }
    
    @discardableResult func visit(_ node:AST)->Int {
        
        if node.name == "BinOp" {
            return self.visit_BinOp(node: node as! BinOp)
        } else if node.name == "Num" {
            return self.visit_Num(node as! Num)
        } else if node.name == "UnaryOp" {
            return self.visit_UnaryOp(node as! UnaryOp)
        } else if node.name == "Var" {
            return self.visit_Var(node as! Var)
        } else if node.name == "Compound" {
            return self.visit_Compound(node as! Compound)
        } else if node.name == "Assign" {
            return self.visit_Assign(node as! Assign)
        }
        
        print("error in visit")
        return 0
    }
    
    func visit_UnaryOp(_ node:UnaryOp)->Int {
        let op = node.token
        if op == .plus {
            return +self.visit(node.expr)
        } else if op == .minus {
            return -self.visit(node.expr)
        }
        print("error in visit")
        return 0
    }
    
    func visit_BinOp(node:BinOp) -> Int {
        if node.token == .plus {
            return self.visit(node.left) + self.visit(node.right)
        } else if node.token == .minus {
            return self.visit(node.left) - self.visit(node.right)
        } else if node.token == .mul {
            return self.visit(node.left) * self.visit(node.right)
        } else if node.token == .div {
            return self.visit(node.left) / self.visit(node.right)
        }
        print("error !!!")
        return 0
    }
    
    func visit_Num(_ node:Num) -> Int {
        return node.value
    }
    
    func visit_NoOp(_ node:NoOp)->Int {
        return -1
    }
    
    func visit_Assign(_ node:Assign)->Int{
        let var_name = node.variable.var_name
        self.global_scope[var_name!] = self.visit(node.expr)
        return -1
    }
    
    func visit_Var(_ node:Var)->Int{
        let var_name = node.var_name
        let val = self.global_scope[var_name!]!
        
        return val
        
    }
    
    func visit_Compound(_ node:Compound)->Int {
        for child in node.children{
            self.visit(child)
        }
        return -1
    }
    
    func interpret()-> Int{
        let tree = self.parser.parse()
        return self.visit(tree)
    }
}

File: Content/Compiler Interpreter/ECHO_LANG/Interpreter.py
from Lexer import *

class NodeVisitor(object):
    def visit(self,node):
        method_name = 'visit_' + type(node).__name__
        visitor = getattr(self,method_name,self.generic_visit)
        return visitor(node)

    def generic_visit(self,node):
        raise Exception('No visit_{} method'.format(type(node).__name__))

class Interpreter(NodeVisitor):
    def __init__(self, parser):
        self.parser = parser
        self.GLOBAL_SCOPE = {}

    def visit_VarDecl(self, node):
        pass

    def visit_Type(self, node):
        pass

    def visit_Var(self, node):
        var_name = node.value
        val = self.GLOBAL_SCOPE[var_name]
        if val is None:
            raise NameError(repr(var_name))
        else:
            return val

    def visit_NoOp(self, node):
        pass

    def visit_BinOp(self, node):
        if node.op.type == PLUS:
            return self.visit(node.left) + self.visit(node.right)
        elif node.op.type == MINUS:
            return self.visit(node.left) - self.visit(node.right)
        elif node.op.type == MUL:
            return self.visit(node.left) * self.visit(node.right)
        elif node.op.type == DIV:
            return self.visit(node.left) / self.visit(node.right)

    def visit_Num(self, node):
        return node.value

    def visit_Assign(self, node):
        var_name = node.left.value
        self.GLOBAL_SCOPE[var_name] = self.visit(node.right)

    def visit_CondBlock(self,node):

        for condBranch in node.condBranchs:
            excuted = self.visit(condBranch) # TODO must has break
            if excuted == 1:
                # mean satisfy condition and excute the block , so quit
                break

    def visit_WhileBlock(self,node):
        cond_node = node.condition
        block = node.block

        while(self.visit(cond_node)) :
            self.visit(block)

    def visit_CondBranchDecl(self,node):
        cond_node = node.condition
        block_node = node.block

        excuted = 0
        isSatisfy = 1
        if cond_node is not None:
            # this is else block
            isSatisfy = self.visit(cond_node)

        if isSatisfy != 0 :
            excuted = 1
            self.visit(block_node)
        return excuted

    def visit_Condition(self,node):
        op = node.op
        leftValue = self.visit(node.left)
        rightValue = self.visit(node.right)

        ret = 0
        if op.type == GREATER:
            ret = leftValue > rightValue
        elif op.type == EQUAL:
            ret = leftValue == rightValue
        elif op.type == LESS:
            ret = leftValue < rightValue

        return ret


    def visit_Program(self, node):
        for statement in node.statements:
            self.visit(statement)

    def interpret(self):
        tree = self.parser.parse()
        return self.visit(tree)